File: BSIGeneralAffairDAL/DALVendor.vb
Imports System.Data
Imports System.Data.SqlClient
Imports BSIGeneralAffairBO
Imports GeneralAffairInterface

Public Class DALVendor
    Implements IVendor

    Public database As DatabaseConfiguration = New DatabaseConfiguration()
    Private dataRead As SqlDataReader
    Public cmd As SqlCommand

    Public Function getAll() As List(Of Vendor) Implements IVendor.getAll
        Dim Vendors As New List(Of Vendor)
        Try
            Dim strSP = "[GeneralAffair].[USP_SelectAllVendor]"
            cmd = New SqlCommand(strSP, database.conn)
            cmd.CommandType = CommandType.StoredProcedure
            database.conn.Open()
            dataRead = cmd.ExecuteReader()
            If dataRead.HasRows Then
                While dataRead.Read
                    Dim Vendor As New Vendor
                    Vendor.VendorID = CInt(dataRead("VendorID"))
                    Vendor.VendorName = dataRead("VendorName").ToString()
                    Vendor.VendorAddress = dataRead("VendorAddress").ToString()
                    Vendor.CreatedAt = dataRead("CreatedAt").ToString()
                    Vendor.UpdatedAt = dataRead("UpdatedAt").ToString()
                    Vendors.Add(Vendor)
                End While
            End If
            dataRead.Close()

            Return Vendors
        Catch ex As Exception
            Throw ex
        Finally
            cmd.Dispose()
            database.conn.Close()
        End Try
    End Function

    Public Function create(obj As Vendor) As Integer Implements IVendor.create
        Try
            Dim strSP = "[GeneralAffair].[USP_StoreVendor]"
            cmd = New SqlCommand(strSP, database.conn)
            cmd.CommandType = CommandType.StoredProcedure

            cmd.Parameters.AddWithValue("@VendorName", obj.VendorName)
            cmd.Parameters.AddWithValue("@VendorAddress", obj.VendorAddress)

            database.conn.Open()
            Dim result = cmd.ExecuteNonQuery()
            If result = 1 Then
                Throw New ArgumentException("Vendor not created")
            End If
            Return result
        Catch sqlex As SqlException
            Throw New ArgumentException(sqlex.Message & " " & sqlex.Number)
        Catch ex As Exception
            Throw ex
        Finally
            cmd.Dispose()
            database.conn.Close()
        End Try
    End Function

    Public Function update(obj As Vendor) As Integer Implements IVendor.update
        Try
            Dim updateVendor = GetById(obj.VendorID)
            If updateVendor Is Nothing Then
                Throw New ArgumentException("Vendor not found")
            End If

            Dim strSP = "[GeneralAffair].[USP_UpdateVendor]"
            cmd = New SqlCommand(strSP, database.conn)
            cmd.CommandType = CommandType.StoredProcedure
            cmd.Parameters.AddWithValue("@VendorID", obj.VendorID)
            cmd.Parameters.AddWithValue("@VendorName", obj.VendorName)
            cmd.Parameters.AddWithValue("@VendorAddress", obj.VendorAddress)

            database.conn.Open()
            Dim result = cmd.ExecuteNonQuery()
            Return result
        Catch sqlex As SqlException
            Throw New ArgumentException(sqlex.Message & " " & sqlex.Number)
        Catch ex As Exception
            Throw ex
        Finally
            cmd.Dispose()
            database.conn.Close()
        End Try
    End Function

    Private Function GetById(vendorID As Integer) As Object
        Try
            cmd = New SqlCommand("SELECT * FROM [GeneralAffair].[Vendors] WHERE VendorID = @VendorID", database.conn)
            cmd.Parameters.AddWithValue("@VendorID", vendorID)
            database.conn.Open()
            dataRead = cmd.ExecuteReader()
            If dataRead.HasRows Then
                dataRead.Read()
                Dim vendor As New Vendor
                vendor.VendorID = CInt(dataRead("VendorID"))
                vendor.VendorName = dataRead("VendorName").ToString
                vendor.VendorAddress = dataRead("VendorAddress").ToString
                vendor.CreatedAt = dataRead("CreatedAt").ToString
                vendor.UpdatedAt = dataRead("UpdatedAt").ToString
                Return vendor
            Else
                Return Nothing
            End If
        Catch ex As Exception
            Throw ex
        Finally
            cmd.Dispose()
            database.conn.Close()
        End Try
    End Function

    Public Function delete(id As Integer) As Integer Implements IVendor.delete
        Try
            Dim updateVendor = GetById(id)
            If updateVendor Is Nothing Then
                Throw New ArgumentException("Vendor not found")
            End If

            Dim strSP = "[GeneralAffair].[USP_DeleteVendor]"
            cmd = New SqlCommand(strSP, database.conn)
            cmd.CommandType = CommandType.StoredProcedure
            cmd.Parameters.AddWithValue("@VendorID", id)

            database.conn.Open()
            Dim result = cmd.ExecuteNonQuery()
        Catch sqlex As SqlException
            Throw New ArgumentException(sqlex.Message & " " & sqlex.Number)
        Catch ex As Exception
            Throw ex
        Finally
            cmd.Dispose()
            database.conn.Close()
        End Try
    End Function
End Class

File: BSIGeneralAffairDAL-C/DALVendor.cs

using BSIGeneralAffairBO_C;
using BSIGeneralAffairDAL_C.Interfaces;
using Dapper;
using System;
using System.Collections.Generic;
using System.Configuration;
using System.Data.SqlClient;
using System.Text;

namespace BSIGeneralAffairDAL_C
{
    public class DALVendor : IVendorDAL
    {
        public void Insert(Vendor entity)
        {
            using (SqlConnection conn = new SqlConnection(GetConnectionString()))
            {
                var strSql = "[GeneralAffair].[USP_StoreVendor]";
                var param = new { 
                    VendorName = entity.VendorName, VendorAddress = entity.VendorAddress 
                };
                try
                {
                    int result = conn.Execute(strSql, param, commandType: System.Data.CommandType.StoredProcedure);
                    if (result == 1)
                    {
                        throw new ArgumentException("Insert data failed..");
                    }
                }
                catch (SqlException sqlEx)
                {
                    throw new ArgumentException($"{sqlEx.InnerException.Message} - {sqlEx.Number}");
                }
                catch (Exception ex)
                {
                    throw new ArgumentException(ex.Message);
                }
            }
        }

        public void Delete(int id)
        {
            using (SqlConnection conn = new SqlConnection(GetConnectionString()))
            {
                var strSql = @"[GeneralAffair].[USP_DeleteVendor]";
                var param = new { VendorID = id };
                try
                {
                    int result = conn.Execute(strSql, param);
                    if (result == 1)
                    {
                        throw new ArgumentException("Delete data failed..");
                    }
                }
                catch (SqlException sqlEx)
                {
                    throw new ArgumentException($"{sqlEx.InnerException.Message} - {sqlEx.Number}");
                }
                catch (Exception ex)
                {
                    throw new ArgumentException(ex.Message);
                }
            }
        }

        public IEnumerable<Vendor> GetAll()
        {
            using (SqlConnection conn = new SqlConnection(GetConnectionString()))
            {
                //var strSql = @"select * from Categories order by CategoryName";
                var strSql = @"[GeneralAffair].[USP_SelectAllVendor]";

                var results = conn.Query<Vendor>(strSql, commandType: System.Data.CommandType.StoredProcedure);
                return results;
            }
        }

        public Vendor GetById(int id)
        {
            using (SqlConnection conn = new SqlConnection(GetConnectionString()))
            {
                var strSql = @"SELECT * FROM [GeneralAffair].[Vendors] Where [VendorID] = @VendorID ";
                var param = new { VendorID = id };
                var result = conn.QuerySingleOrDefault<Vendor>(strSql, param);
                return result;
            }
        }

        public IEnumerable<Vendor> GetByName(string name)
        {
            using (SqlConnection conn = new SqlConnection(GetConnectionString()))
            {
                var strSql = @"SELECT * FROM [GeneralAffair].[Vendors] Where [VendorName] like @VendorName ";
                var param = new { VendorName = $"%{name}%" };
                var result = conn.Query<Vendor>(strSql, param);
                return result;
            }
        }

        public void Update(Vendor entity)
        {
            using (SqlConnection conn = new SqlConnection(GetConnectionString()))
            {
                try
                {
                    var strSql = @"[GeneralAffair].[USP_UpdateVendor]";
                    var param = new { VendorID = entity.VendorID, VendorName = entity.VendorName, VendorAddress = entity.VendorAddress };
                    int result = conn.Execute(strSql, param);

                    //jika result = -1, berarti update data gagal
                    if (result == 1)
                    {
                        throw new Exception("Update data failed..");
                    }
                }
                catch (SqlException sqlEx)
                {
                    throw new Exception($"{sqlEx.Message} - {sqlEx.Number}");
                }
                catch (Exception ex)
                {
                    throw new Exception(ex.Message);
                }
            }
        }

        private string GetConnectionString()
        {
            return Helper.GetConnectionString();
            //return @"Data Source=ACTUAL;Initial Catalog=LatihanDb;Integrated Security=True;TrustServerCertificate=True";
            //return ConfigurationManager.ConnectionStrings["MyDbConnectionString"].ConnectionString;
        }

        public int GetCountVendors(string name)
        {
            using (SqlConnection conn = new SqlConnection(GetConnectionString()))
            {
                var strSql = @"SELECT COUNT(*) FROM [GeneralAffair].[Vendors] Where [VendorName] like @VendorName ";
                var param = new { VendorName = $"%{name}%" };
                var result = Convert.ToInt32(conn.ExecuteScalar(strSql, param));
                return result;
            }
        }

        public IEnumerable<Vendor> GetWithPaging(int pageNumber, int pageSize, string name)
        {
            using (SqlConnection conn = new SqlConnection(GetConnectionString()))
            {
                var strSql = @"SELECT COUNT(*) FROM [GeneralAffair].[Vendors] 
                              Where [VendorName] like @VendorName 
                              order by VendorName OFFSET @Offset ROWS FETCH NEXT @PageSize ROWS ONLY";
                var param = new { VendorName = $"%{name}%", Offset = (pageNumber - 1) * pageSize, PageSize = pageSize };
                var results = conn.Query<Vendor>(strSql, param);
                return results;
            }
        }
    }
}
